Automatic Recognition of Guesser UTXOs

aszepieniec opened the pull request:
The guesser UTXOs are locked through a `RawHashLock`. Only the guesser is knowledgeable of the key that can release this lock.

Prior to this PR, this key was already deterministically derived from the wallet secret and the previous block hash. Despite this deterministic derivation, the rest of the pipeline assumed no such structure. In particular, a successful guesser task would pass the guesser preimage upstream; and explicit `ExpectedUtxo`s would be generated so that the wallet can claim these UTXOs.

This commit
 - Modifies the wallet update function so that it checks an incoming block against what the guesser preimage would be if this wallet had been the one to guess it. This change obviates the need to transmit `ExpectedUtxo`s to the wallet update function.
 - Removes the parts of the mechanism whereby an unstructured guesser preimage or `ExpectedUtxo`s for guesser fee UTXOs are passed around.
 
Also:
 - Refactor method `is_timelocked_but_otherwise_spendable_at` on `Utxo`. While this refactor was useful in debugging this PR, it is tangentially if at all related.

Closes #413.

bittoshimoto commented:
Tested and works perfectly !

2025-02-15T09:27:10.559299132Z  INFO ThreadId(01) neptune_cash::models::state::wallet::wallet_state: Received UTXO in block 02305052090323280570,02498409249708683187,16445942019016897052,07935533845435965986,00000000000062517739, height 142
value = 33.28000000
is guesser fee: true

dan-da reviewed (commented):
I had a node with a corrupted wallet database that had previously mined (guessed) a testnet block.  I had previously removed the wallet db and blockchain, and performed a full resync.  However, that did not detect the guesser utxos, so the wallet balance remained 0.

As a test of this PR, I moved the wallet DB away, leaving the other databases and wallet files.  I then started neptune-core with this branch applied.  The guesser utxos were detected right away, and the wallet balance and history appear correct.

As such, this PR seems a definite improvement.   And we could probably merge it, at least as a fix for now.

----------------

That said, I have serious reservations about the entire design of guessing keys and treating guessed utxos differently from every other utxo, including composer rewards.   quoting from a review comment:

> composers and guessers share the same coinbase reward. It is fundamentally just two separate utxos that needs to be claimed by each party.
> 
> It seems strange and non-symmetric to me that they use different key types, different notification code paths, and different ways to restore them.
> 
> I think it is a design-smell, and that smell may be telling us to re-think.

I confess I do not understand the guesser-key implementation in fullness.  So perhaps I should not render any opinion.   But I will anyway.  Here are some things that bother me:

1. composer coinbase utxos and guesser coinbase utxos are now treated differently, for reasons I don't understand, as per above.
2. guesser utxos use a unique key type, which also seems strange and non-symmetric to me.  In particular, it is derived from *both* the wallet seed and the block digest.   As such, given only a wallet seed, one cannot derive these keys.   This would break any 3rd party key derivation tools.
3. This new key type has added complexity to the code, in particular the SpendingKey and  ReceivingAddress abstractions were abused in ways not intended in their design.
4. These changes were made in a rush without adequate planning or testing of what happens when a wallet needs to restore from seed.  This should *always* be a key consideration for us when making changes.
5. The key type is displayed differently from other keys in the dashboard addresses screen.  There is no unique identifier.  It's another example of the key seeming "special" and not in harmony with the other key types.
6. This seems like complexity for the sake of complexity's sake.  I do not understand what problem it is solving that could not be solved with the existing mechanisms.

Ok.  those are my thoughts.  admittedly from a place of incomplete understanding.   Happy to entertain counter-arguments, and knowledge dumps. 

--------

Given above design concerns, I am not marking this PR approved.   Nor requesting changes.

If we were to come to a plan for deprecating this key type (or making it derivable from wallet seed *only*) and treat composer utxos and guesser utxos the same, then I would be more comfortable approving this PR as a temporary measure, with understanding that we have a plan/roadmap towards something less complex and more elegant.

aszepieniec commented:
>  So perhaps I should not render any opinion.

I disagree with this sentiment. Articulating a premature opinion is a great way to a) crystallize the source of the confusion; and b) invite counter-argument. Both processes accelerate the evolution towards a well formed opinion. I would caution against setting premature opinions in stone, but that's different from articulating them.

For the record and for completeness of context, let me repeat something I wrote elsewhere:

> We did not add a `RawHashLock` address type; we added a `RawHashLock` key type. Guesser fees are like mining rewards (and often are exactly that). They don't originate from somewhere, and you can't send them anywhere (by which I mean: make them the output of a transaction). What you can do, is unlock them (to spend them, or to make them the input of a transaction). The type system should reflect that asymmetry.
>
> RawHashLock did come about through a historical meander: originally the nonce had a preimage and that preimage was the key to unlock the guesser fee for spending. In the tight guessing loop, you want to hash as few times as possible and avoid any overhead that comes with key derivation, encrypting, or address formatting. `SymmetricKey` is inadequate in this context because of this overhead. So that's where `RawHashLock` was born.
>
> Then we switched away from using nonce-preimages and instead added the extra field on the block header. Now the guesser key is prepared before the tight guesser loop starts. In principle `SymmetricKey` can do the job; the overhead is negligible now. But we already had `RawHashLock` lying around and that's a better fit of semantics onto types because of the asymmetry.

Now on to your points.

> composers and guessers share the same coinbase reward. It is fundamentally just two separate utxos that needs to be claimed by each party.
>
> It seems strange and non-symmetric to me that they use different key types, different notification code paths, and different ways to restore them.

Here is a relevant difference between composer fees and guesser fees: The composer has complete control over the exact shape of the transaction that is confirmed by a block, so long as it is valid and confirmable. The composer literally starts with a coinbase transaction, which is the only type of transaction that a) unmerged, and b) allowed to inflate the money supply. The composer can specify any number of destination UTXOs, any set of unlock policies, and even include arbitrary coin types. The only requirement is that at least half of the new money supply is time-locked for 3 years.

By contrast, the guesser has almost no control over the guesser UTXOs. He *only* supplies a hash digest. The guesser UTXOs are defined implicitly by the block; they are not listed among the transaction's outputs. This is why method `mutator_set_after()` exists -- it accounts for two more UTXOs than the ones you can read from the block's transaction. The guesser UTXOs consists of one native currency coin, and one native currency coin with timelock coin, respectively. The guesser cannot add other coins. The lock script is a (in the context of proving its correct execution) a proof of preimage knowledge and the guesser cannot specify any other options. It happens automatically.

> 1. composer coinbase utxos and guesser coinbase utxos are now treated differently, for reasons I don't understand, as per above.

It is not just the wallet that treats them differently; it is also the blockchain's consensus rules. The consensus rules specify what the guesser fee UTXO has to look like, but imposes no constraints on composer fees.

> 2. guesser utxos use a unique key type, which also seems strange and non-symmetric to me. In particular, it is derived from both the wallet seed and the block digest. As such, given only a wallet seed, one cannot derive these keys. This would be incompatible with any 3rd party key derivation tools including hardware wallets.

Assuming you trust that third party, then the requirement that they (the 3rd party) also have access to block hashes is not a big deal. It is public information and not a lot of data which by the way can be streamed. Moreover, deriving the key from the secret seed and the block digest is preferable to deriving it from the secret seed and a counter. The counter, after all, represents state which can be corrupted or lost or inconsistent across multiple machines that should be in sync. Note that this PR, which automagically tracks own guesser UTXOs, was a breeze to implement because of this statelessness; whereas I expect the followup, which automagically tracks change UTXOs, to be much more challenging. (Give it a try if you don't share my expectation ;-])

If you do *not* trust that third party, then you have a point -- but change UTXOs are vulnerable to the same criticism. You can generate a bunch of symmetric encryption keys that the 3rd party can then use to trial decrypt public announcements, but then you are still doing a linear amount of work and defeating the purpose of outsourcing it. If instead you send them the secret seed then they can derive the symmetric encryption keys but they can also spend any change UTXOs that they find. Moreover miners are already running heavy-duty machinery; support for light-weight clients capable of securely outsourcing work is targeting a different audience from miners.

> 3. This new key type has added complexity to the code, in particular the SpendingKey and ReceivingAddress abstractions were abused in ways not intended in their design.

I'm not sure this proposition counts as an argument even if we take it at face value. Designs are allowed to evolve in response to a better understanding of the problems they are supposed to solve. When you wrote the `AddressType` and `KeyType` abstractions you did not have guesser fee UTXOs on your mind -- and neither did I when I reviewed it. I'm not sure you agree but in my mind a guesser fee UTXO is completely different from a change UTXO.

Digging deeper, is there some particular documentation, or other articulation of the intended design, that the modifications since then are at odds with? 

> 4. These changes were made in a rush without adequate planning or testing of what happens when a wallet needs to restore from seed.

There is no doubt that the code in question was written under time pressure. That said, we (myself and Thorkil) wrote this together between lots of oral discussions of which none made it to any written record. Perhaps the absence of record of these discussions, or your (unintentional!) exclusion from them, paints an inaccurate picture of sloppiness and degrading standards.

> This [recovery from seed] should always be a key consideration for us when making changes.

To avoid confusion, we certainly did not change any existing mechanisms for recovering UTXOs from a seed.

> 5. The key type is displayed differently from other keys in the dashboard addresses screen. There is no unique identifier. It's another example of the key seeming "special" and not in harmony with the other key types.

This is a superficial cosmetic. Feel free to add unique identifiers, for instance by displaying the after-image, or by displaying the block height and hash, or something else entirely.

> 6. This seems like complexity for the sake of complexity's sake. I do not understand what problem it is solving that could not be solved with the existing mechanisms.

Would you mind explaining from a high level perspective how the existing mechanisms could:
 - effect ownership by the winning guesser of the guesser fee UTXOs;
 - enable that guesser to recover guesser fee UTXOs from seed?

I'm not saying it's impossible. But in order to have a fruitful discussion I cannot be guessing at the alternative.

For the sake of improving your understanding and furthering the discussion, you are allowed to ignore consensus rules. However, the constraints imposed by the existing consensus mechanism become relevant again if you want to articulate a proposal for moving forwards without a fork.

dan-da commented:
> If you do _not_ trust that third party, then you have a point -- but change UTXOs are vulnerable to the same criticism. You can generate a bunch of symmetric encryption keys that the 3rd party can then use to trial decrypt public announcements, but then you are still doing a linear amount of work and defeating the purpose of outsourcing it. If instead you send them the secret seed then they can derive the symmetric encryption keys but they can also spend any change UTXOs that they find. Moreover miners are already running heavy-duty machinery; support for light-weight clients capable of securely outsourcing work is targeting a different audience from miners.

I think we are talking about different things here.  I am not talking about a 3rd party entity deriving keys... that's a huge risk with any key type.   I said "This would be incompatible with any 3rd party key derivation **tools** including hardware wallets."   By this, I mean any software or hardware *tool* that is not part of neptune-core, and/or does not have access to a full-node chain.  Such a tool would be operated by the wallet owner.

In particular I do not see how a hardware wallet would derive these keys.    What is your answer to that?

(afaik, present day hardware wallets simply generate (or accept) a wallet seed and then deterministically derive keys from that.)

aszepieniec commented:
> > If you do _not_ trust that third party, then you have a point -- but change UTXOs are vulnerable to the same criticism. You can generate a bunch of symmetric encryption keys that the 3rd party can then use to trial decrypt public announcements, but then you are still doing a linear amount of work and defeating the purpose of outsourcing it. If instead you send them the secret seed then they can derive the symmetric encryption keys but they can also spend any change UTXOs that they find. Moreover miners are already running heavy-duty machinery; support for light-weight clients capable of securely outsourcing work is targeting a different audience from miners.
> 
> I think we are talking about different things here. I am not talking about a 3rd party entity deriving keys... that's a huge risk with any key type. I said "This would be incompatible with any 3rd party key derivation **tools** including hardware wallets." By this, I mean any software or hardware _tool_ that is not part of neptune-core, and/or does not have access to a full-node chain. Such a tool would be operated by the wallet owner.
> 
> In particular I do not see how a hardware wallet would derive these keys. What is your answer to that?
> 
> (afaik, present day hardware wallets simply generate (or accept) a wallet seed and then deterministically derive keys from that.)

Okay, I obviously don't understand, so let me ask a basic question.

The use case in question is that of a third-party key derivation tool deriving keys without access to blockchain information. Why is this use case worth supporting? Third party key derivation tools are only useful in conjunction with blockchain information. Whether it is about initiating transactions, or about computing and showing balance evolution over time, or even auditing solvency or producing a proof thereof, it seems like *some* blockchain data is always necessary to give context to the derived keys. Otherwise you have a device that derives keys but is useless for anything besides that.

dan-da commented:
> Otherwise you have a device that derives keys but is useless for anything besides that

that's pretty much the point of a hardware wallet, as I understand it.  It is a simple device that just derives keys on demand.  The seed is only stored on the hardware wallet and not on the device running the wallet-software.  When the wallet software needs a key, it asks the hardware wallet for the key at the corresponding derivation index.    That's it.  to be clear: the hardware wallet never receives or stores any blockchain data.

afaik, this is how hardware wallets have been working industry wide for many years now.  It is possible that they do "smarter" things these days.   Or my understanding could be wrong.  So I am open to information showing otherwise.

LIkewise, I and others have written software tools that "just" derive keys.   These are useful in all kinds of situations when developing, debugging, looking for lost funds, etc.   With transparent blockchains an obvious use-case is to generate addresses and then check a block explorer to see if any of them have received funds.  Another use case is to generate a receiving address without access to a running wallet/node.  It is a very nice property that given only a wallet seed, a lightweight tool can generate all possible keys for a wallet, without requiring any blockchain data.

I hope you can see that this property of "derive all keys from wallet seed" has been an industry standard since approx 2015 that people are accustomed to when it comes to wallet management.  Deviating from that is kind of a big deal.  I think it will take some time to wrap my mind around, and others as well, as neptune community grows.  Maybe these guessing keys are a really nice innovation...   I just don't think any of us understand all implications yet (positive or negative). 

ok, so that said.  let's return to how a hardware wallet can be implemented.  now that I understand these `HashLockKey` a _little bit_ better, it seems straight-forward.  A neptune hardware wallet could be developed with a derive-guesser-key() API that would accept a block-digest as argument.    It would be pretty much this fn from WalletSecret, running on the hardware wallet:

```
    pub(crate) fn guesser_spending_key(&self, prev_block_digest: Digest) -> HashLockKey {
        HashLockKey::from_preimage(Tip5::hash_varlen(
            &[
                self.secret_seed.0.encode(),
                vec![hash_lock_key::RAW_HASH_LOCK_KEY_FLAG],
                prev_block_digest.encode(),
            ]
            .concat(),
        ))
    }
```
where the hardware wallet provides the secret_seed and the software wallet provides the prev_block_digest.

--------

So it seems a hardware wallet (or other 3rd party tool) can work with a software wallet, to the extent the software wallet has historical block hashes.  But it cannot generate a sequence of keys by itself without blockchain data.  Therein is the deviation from industry practices (again, for better or worse).

thinking about implications:

It appears to me that this poses a challenge for light nodes.  Imagine a miner that mined a bunch of blocks in the early days and years later restores seed to a light node, and is surprised to not find the guessing rewards.

Now it may be possible (or planned?) that a light node will request all historic block headers from full nodes.  The headers being *much* smaller than the bodies.  In which case, I think the problem is resolved, yes?   (At least the key derivation aspect of it.)

aszepieniec commented:
> > Otherwise you have a device that derives keys but is useless for anything besides that
> 
> that's pretty much the point of a hardware wallet, as I understand it. It is a simple device that just derives keys on demand. The seed is only stored on the hardware wallet and not on the device running the wallet-software. When the wallet software needs a key, it asks the hardware wallet for the key at the corresponding derivation index. That's it. to be clear: the hardware wallet never receives or stores any blockchain data.
> 
> afaik, this is how hardware wallets have been working industry wide for many years now. It is possible that they do "smarter" things these days. Or my understanding could be wrong. So I am open to information showing otherwise.
> 
> LIkewise, I and others have written software tools that "just" derive keys. These are useful in all kinds of situations when developing, debugging, looking for lost funds, etc. With transparent blockchains an obvious use-case is to generate addresses and then check a block explorer to see if any of them have received funds. Another use case is to generate a receiving address without access to a running wallet/node. It is a very nice property that given only a wallet seed, a lightweight tool can generate all possible keys for a wallet, without requiring any blockchain data.
> 
> I hope you can see that this property of "derive all keys from wallet seed" has been an industry standard since approx 2015 that people are accustomed to when it comes to wallet management. Deviating from that is kind of a big deal. I think it will take some time to wrap my mind around, and others as well, as neptune community grows. Maybe these guessing keys are a really nice innovation... I just don't think any of us understand all implications yet (positive or negative).

I guess I am a little confused about this particular question: where does the public key cryptography happen?

 - a) in the wallet software, not on the hardware device. Frankly, this configuration makes very little sense to me. It is technically possible to generate individual keys without leaking master key material, but if all these individual keys are read by the same wallet software, it defeats the point of using a hardware device to begin with. With a little strain I can imagine a highly complex and hierarchical organization with high-level managers upon mid- and low-level managers where this kind of configuration makes sense; for casual individuals I don't think it does.
 - b) on the hardware device, not in the wallet software. This configuration makes perfect sense to me. The wallet software only sees public keys, but that's enough to do bookkeeping. If the user wants to send a transaction, the wallet software prepares it and ultimately the hardware device signs it.
 - c) both on the hardware device and in the wallet software. I can see how this configuration inherits the convenience of both prior options, but I don't see how it inherits the security guarantees (arguably the main selling point) of option (b).

Since you have more experience in this field than I do, maybe you can weigh in on which option is the industry standard, or correct my understanding at a more fundamental level. The text I quote above makes perfect sense to me if configuration (b) applies. Pending said correction, I put forward the following argument, which assumes (b) is the industry standard configuration and the right-hand-side of the implicit comparison. 

The relevant difference between Neptune Cash and all (?) other cryptocurrencies (certainly all that I am aware of) is that in Neptune Cash that there are no public keys. (There are public *encryption* keys but these are optional and not (or at best, indirectly) used for transferring ownership -- I mean *signature* public keys.) Consequently, configuration (b) simply does not apply. The other two don't either: the public key operations happen no-where.

Since public keys cannot be derived on hardware devices, allowing the hardware device to derive keys (which are symmetric) approximates configuration (a), not configuration (b).

The deviation from standard practice is an artifact not of guesser keys but of the absence of public keys. In fact, our existing symmetric key and generation address infrastructure already deviates from standard practice.

> ok, so that said. let's return to how a hardware wallet can be implemented. now that I understand these `HashLockKey` a _little bit_ better, it seems straight-forward. A neptune hardware wallet could be developed with a derive-guesser-key() API that would accept a block-digest as argument. It would be pretty much this fn from WalletSecret, running on the hardware wallet:
> 
> ```
>     pub(crate) fn guesser_spending_key(&self, prev_block_digest: Digest) -> HashLockKey {
>         HashLockKey::from_preimage(Tip5::hash_varlen(
>             &[
>                 self.secret_seed.0.encode(),
>                 vec![hash_lock_key::RAW_HASH_LOCK_KEY_FLAG],
>                 prev_block_digest.encode(),
>             ]
>             .concat(),
>         ))
>     }
> ```
> 
> where the hardware wallet provides the secret_seed and the software wallet provides the prev_block_digest.

Yes. Agreed.

> It appears to me that this poses a challenge for light nodes. Imagine a miner that mined a bunch of blocks in the early days and years later restores seed to a light node, and is surprised to not find the guessing rewards.
> 
> Now it may be possible (or planned?) that a light node will request all historic block headers from full nodes. The headers being _much_ smaller than the bodies. In which case, I think the problem is resolved, yes? (At least the key derivation aspect of it.)

Extending the peer-to-peer protocol with messages enabling nodes to ask their peers for historical headers is definitely an option, and I do think that solves the problem of deriving guesser keys for guessed blocks.

Side note: I try to distinguish between full and archival nodes as follows:
 - full node: one that is capable of deciding canonicity between new competing forks without relying on peers for supplementary or historical data;
 - archival node: one that stores all historical data.
Note that deciding canonicity is not something you can do with the block headers alone because that only tells you which fork is the heaviest and you *also* need to know it is valid. So in particular, in the context of Bitcoin, full nodes and archival nodes denote the same thing; but in the context of succinct chains (which Neptune Cash will one day be) full nodes can be different from archival nodes. And you would be querying *archival* nodes for historical blockchain data.

Sword-Smith commented on src/main_loop.rs:
```diff
@@ -517,13 +517,9 @@ impl MainLoopHandler {
                         new_block.as_ref().clone(),
-                        [
-                            new_block_info.composer_utxos,
-                            new_block_info.guesser_fee_utxo_infos,
```
> composers and guessers share the same coinbase reward. It is fundamentally just two separate utxos that needs to be claimed by each party.

The cryptographic data in composer rewards are known prior to the construction of the coinbase transaction, because they only depend on data from the current tip. The cryptographic data in the guesser rewards, however, is only known *after* the PoW puzzle has been solved for the next block.

That means that the two types of rewards need to use different code paths, since an `ExpectedUtxo` object cannot be created for guesser rewards -- as the `sender_randomness` of the guesser rewards is the hash of the block and is thus not known prior to the solution of the PoW puzzle.

aszepieniec commented:
Merged into master, but by all means let's continue the discussion.
